src/main/java/org/uast/astgen/analyzer/Analyzer.java

```diff
@@ -0,0 +1,528 @@
+/*
+ * MIT License Copyright (c) 2022 unified-ast
+ * https://github.com/unified-ast/ast-generator/blob/master/LICENSE.txt
+ */
+
+package org.uast.astgen.analyzer;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.Stack;
+import java.util.stream.Collectors;
+import org.uast.astgen.exceptions.ExpectedOnlyOneEntity;
+import org.uast.astgen.rules.Child;
+import org.uast.astgen.rules.Descriptor;
+import org.uast.astgen.rules.Disjunction;
+import org.uast.astgen.rules.Node;
+import org.uast.astgen.rules.Statement;
+
+/**
+ * Analyzes nodes hierarchy described with DSL.
+ *
+ * @since 1.0
+ */
+@SuppressWarnings("PMD.CloseResource")
+public class Analyzer {
+    /**
+     * Nodes related to current programming language and green nodes.
+     */
+    private final List<Node> nodes;
+
+    /**
+     * The programming language for which the analysis is conducted.
+     */
+    private final String language;
+
+    /**
+     * Mappings between a node type and a result of its analysis.
+     */
+    private final Map<String, Result> info;
+
+    /**
+     * Stack to save nodes with tagged types to be analyzed.
+     */
+    private final Stack<String> stack;
+
+    /**
+     * Constructor.
+     * @param descriptors The node descriptors
+     * @param language The programming language
+     * @throws ExpectedOnlyOneEntity exception if nodes described in rules
+     *  contain duplications
+     */
+    public Analyzer(
+        final List<Statement<Node>> descriptors,
+        final String language) throws ExpectedOnlyOneEntity {
```

Comment:
Please create another exception for this case, say, 'DuplicateRule'

Reply:
It should extend the 'GeneratorException' class.

Reply:
Done.